examples/slackbot/Notebook.ipynb

```diff
@@ -74,120 +67,89 @@
+    "    duration = k.minutes(5)  # !!!\n",
+    "\n",
+    "    prompts = build_prompts(messages)\n",
+    "    shifted_prompts = prompts.shift_by(duration)  # !!!\n",
     "\n",
-    "    k.record({\n",
-    "        # A list of all messages over the past 10 minutes (up to 100)\n",
-    "        \"recent_messages\": messages\n",
-    "            #.select(\"user\", \"type\", \"text\")\n",
-    "            .select(False, \"user\", \"subtype\", \"text\")\n",
-    "            #.collect(window=since(conversation_start), max=100),\n",
-    "            .last(),\n",
+    "    reaction_users = prompts.col(\"reactions\").col(\"name\").collect(w.trailing(duration)).explode()  # !!!\n",
+    "    participating_users = prompts.col(\"user\").collect(w.trailing(duration))  # !!!\n",
+    "    engaged_users = k.union(reaction_users, participating_users)  # !!!\n",
     "\n",
-    "        # How many messages have been reacted to in the conversation\n",
-    "        \"reaction_count\": messages\n",
-    "            #.filter(messages[\"reactions\"].is_not_null())\n",
-    "            #.count(window=since(conversation_start)),\n",
-    "            [\"reply_count\"].sum(),\n",
-    "    })"
+    "    return k.record({ \"prompt\": shifted_prompts, \"completion\": engaged_users}) \\\n",
+    "        .filter(shifted_prompts.is_not_null())  # !!!"
    ]
   },
   {
-   "cell_type": "code",
-   "execution_count": null,
-   "id": "fdb2d959-d371-4026-9f8d-4ab26cfbf317",
+   "cell_type": "markdown",
+   "id": "0035f558-23bd-4b4d-95a0-ed5e8fece673",
    "metadata": {},
-   "outputs": [],
    "source": [
-    "def examples(messages):\n",
-    "    # We'll train ChatGPT to generate the user ID who will engage next\n",
-    "    k.record({\n",
-    "        # For each example, use the previous prompt\n",
-    "        \"prompt\": prompt(messages)\n",
-    "            #.lag(1),\n",
-    "            .last(),\n",
-    "\n",
-    "        # ...and the current user ID\n",
-    "        \"completion\": messages[\"user\"],\n",
-    "    })"
+    "## Fine-tune the model"
    ]
   },
   {
    "cell_type": "code",
    "execution_count": null,
-   "id": "0be6bde5-9602-43e6-89bf-e0e1470fc7c0",
+   "id": "af7d2a45-eb89-47ce-b471-a39ad8c7bbc7",
    "metadata": {},
    "outputs": [],
    "source": [
-    "# Format the data for OpenAI\n",
-    "def format_prompt(prompt):\n",
-    "    return SYSTEM_CONTEXT + json.dumps(prompt) + \"\\n\\n###\\n\\n\"\n",
-    "def format_completion(completion):\n",
-    "    return completion + \"###\""
-   ]
-  },
-  {
-   "cell_type": "markdown",
-   "id": "0035f558-23bd-4b4d-95a0-ed5e8fece673",
-   "metadata": {},
-   "source": [
-    "## Fine-tune the model"
+    "import pandas\n",
+    "\n",
+    "input_df = pandas.read_parquet(\"./messages.parquet\")\n",
+    "messages = sources.ArrowSource(\"ts\", \"channel\", input_df)\n",
+    "messages = messages.with_key(k.record({  # !!!\n",
+    "        \"channel\": messages.col(\"channel\"),\n",
```

Comment:
We don't have a `col` method. Did you mean `messages["thread_id"]`?

Reply:
Although, maybe it would be better to use `col` *exclusively* for field access, and have `base[index]` be just for list/map indexing. One, that let's those be more clearly distinct. Two, it would allow `col` to automatically apply to `list<record>` and `map<_, record>` by applying `col` across the records.

Reply:
Yeah, I think the issue is that `[]` is ambiguous if you have a map of records.